python/setup.py

```diff
@@ -14,80 +14,132 @@
 # limitations under the License.
 #
 
+from Cython.Build import cythonize
+from distutils.sysconfig import get_python_lib
 from setuptools import setup, find_packages
 from setuptools.extension import Extension
-from Cython.Build import cythonize
+from setuputils import get_submodule_dependencies
+
 import os
-import versioneer
-from distutils.sysconfig import get_python_lib
+import subprocess
 import sys
+import sysconfig
+import versioneer
+import warnings
 
 install_requires = [
-    'numpy',
+    'numba',
     'cython'
 ]
 
-cuda_include_dir = '/usr/local/cuda/include'
-cuda_lib_dir = "/usr/local/cuda/lib"
 
-if os.environ.get('CUDA_HOME', False):
-    cuda_lib_dir = os.path.join(os.environ.get('CUDA_HOME'), 'lib64')
-    cuda_include_dir = os.path.join(os.environ.get('CUDA_HOME'), 'include')
+##############################################################################
+# - Dependencies include and lib folder setup --------------------------------
 
+CUDA_HOME = os.environ.get("CUDA_HOME", False)
+if not CUDA_HOME:
+    CUDA_HOME = (
+        os.popen('echo "$(dirname $(dirname $(which nvcc)))"').read().strip()
+    )
+cuda_include_dir = os.path.join(CUDA_HOME, "include")
+cuda_lib_dir = os.path.join(CUDA_HOME, "lib64")
 
-rmm_include_dir = '/include'
-rmm_lib_dir = '/lib'
 
-# including cumlprims dirs, in case rmm paths change in the future
-cumlprims_include_dir = '/include'
-cumlprims_lib_dir = '/lib'
+##############################################################################
+# - Subrepo checking and cloning ---------------------------------------------
 
-if os.environ.get('CONDA_PREFIX', None):
-    conda_prefix = os.environ.get('CONDA_PREFIX')
-    rmm_include_dir = conda_prefix + rmm_include_dir
-    rmm_lib_dir = conda_prefix + rmm_lib_dir
+subrepos = [
+    'cub',
+    'cutlass',
+    'faiss',
+    'treelite'
+]
 
-    cumlprims_include_dir = conda_prefix + cumlprims_include_dir
-    cumlprims_lib_dir = conda_prefix + cumlprims_lib_dir
+# We check if there is a libcuml++ build folder, by default in cpp/build
+# or in CUML_BUILD_PATH env variable. Otherwise setup.py will clone the
+# dependencies defined in cpp/CMakeListst.txt
+if "clean" not in sys.argv:
+    if os.environ.get('CUML_BUILD_PATH', False):
+        libcuml_path = '../' + os.environ.get('CUML_BUILD_PATH')
+    else:
+        libcuml_path = '../cpp/build/'
+
+    found_cmake_repos = get_submodule_dependencies(subrepos,
+                                                   libcuml_path=libcuml_path)
+
+    if found_cmake_repos:
+        treelite_path = libcuml_path + 'treelite/src/treelite'
```

Comment:
Not urgent here, but for future upgrades, I really like using os.path.join instead of concating paths because I'm always paranoid about having to remember whether the first part has a trailing slash or not.